File: semestre-1/disciplina-3-arquitetura-de-dados/unidade-2/aula-2/aula-2.py
# aula-2.py
# Exemplo em Python: Simulação de Modelos Lógicos e Áreas de Interesse em Varejo.
# Usamos classes (POO) para representar as entidades (MLC) e um sistema de agregação
# para simular as Áreas de Interesse e os Repositórios Lógicos.

# --- 1. MODELO LÓGICO CORPORATIVO (MLC): As Entidades e seus Atributos ---

class Cliente:
    """Representa a entidade Cliente do MLC."""
    def __init__(self, id_cliente, nome, regiao):
        self.id_cliente = id_cliente  # Atributo: Identificação
        self.nome = nome              # Atributo: Nome do cliente
        self.regiao = regiao          # Atributo: Região de moradia

    def __repr__(self):
        return f"Cliente(ID: {self.id_cliente}, Nome: {self.nome}, Região: {self.regiao})"

class Produto:
    """Representa a entidade Produto do MLC."""
    def __init__(self, id_produto, nome, preco):
        self.id_produto = id_produto # Atributo: Identificação
        self.nome = nome             # Atributo: Nome do produto
        self.preco = preco           # Atributo: Preço unitário

    def __repr__(self):
        return f"Produto(ID: {self.id_produto}, Nome: {self.nome}, Preço: R${self.preco:.2f})"

class Venda:
    """Representa a entidade Transação de Venda (o relacionamento entre Cliente e Produto)."""
    def __init__(self, id_venda, id_cliente, id_produto, quantidade, data):
        self.id_venda = id_venda
        self.id_cliente = id_cliente
        self.id_produto = id_produto
        self.quantidade = quantidade
        self.data = data

    def __repr__(self):
        return f"Venda(ID: {self.id_venda}, Cliente: {self.id_cliente}, Produto: {self.id_produto}, Qtd: {self.quantidade}, Data: {self.data})"

# --- 2. MODELO DE ÁREA DE INTERESSE (MAI) & MODELO LÓGICO DE REPOSITÓRIO (MLR) ---

class DataMartMarketing:
    """
    Simula um Data Mart de Marketing, uma Área de Interesse.
    Foca apenas em dados de Cliente e Vendas para segmentação e campanhas.
    """
    def __init__(self):
        # O repositório armazena listas (simulando tabelas filtradas)
        self.dados_clientes = {}
        self.dados_vendas = []
        print("\n[MAI/MLR] Data Mart de Marketing inicializado.")

    def carregar_dados_brutos(self, clientes, vendas):
        """Carrega e filtra dados relevantes para Marketing."""
        # Marketing está interessado em Clientes (para segmentação) e Vendas (para comportamento)
        for cliente in clientes:
            self.dados_clientes[cliente.id_cliente] = cliente
        
        # Simula a agregação/carregamento no repositório
        for venda in vendas:
            self.dados_vendas.append(venda)
        print(f"[Marketing] Carregados {len(self.dados_clientes)} clientes e {len(self.dados_vendas)} vendas.")

    def analisar_clientes_por_regiao(self, regiao):
        """Exemplo de análise que o time de Marketing faria."""
        clientes_segmentados = [
            c for c in self.dados_clientes.values() if c.regiao == regiao
        ]
        return clientes_segmentados

class DataWarehouseVendas:
    """
    Simula um Data Warehouse (DW), uma Área de Interesse/Repositório central.
    Armazena todos os fatos (Vendas) e dimensões (Cliente, Produto) para análise histórica.
    """
    def __init__(self):
        self.fato_vendas = []
        self.dim_clientes = {}
        self.dim_produtos = {}
        print("\n[MAI/MLR] Data Warehouse de Vendas inicializado.")

    def carregar_dados_completos(self, clientes, produtos, vendas):
        """Carrega todos os dados (entidades) no repositório DW."""
        for c in clientes:
            self.dim_clientes[c.id_cliente] = c
        for p in produtos:
            self.dim_produtos[p.id_produto] = p
        
        # Fato Vendas
        self.fato_vendas.extend(vendas)
        print(f"[DW Vendas] Carregados {len(self.fato_vendas)} registros de vendas para análise.")

    def calcular_faturamento_total(self):
        """Exemplo de métrica de negócio obtida a partir do modelo lógico (fatos e dimensões)."""
        faturamento = 0
        for venda in self.fato_vendas:
            produto = self.dim_produtos.get(venda.id_produto)
            if produto:
                faturamento += venda.quantidade * produto.preco
        return faturamento

# --- DADOS DE EXEMPLO (Simulação da Coleta) ---

clientes_dados = [
    Cliente(101, "Ana Silva", "Sul"),
    Cliente(102, "Bruno Costa", "Norte"),
    Cliente(103, "Carla Meirelles", "Sul"),
    Cliente(104, "David Lopes", "Norte"),
]

produtos_dados = [
    Produto(501, "Celular X", 1500.00),
    Produto(502, "Notebook Y", 4500.00),
    Produto(503, "Fone Z", 250.00),
]

vendas_dados = [
    Venda(1, 101, 501, 1, "2024-09-01"),
    Venda(2, 102, 503, 2, "2024-09-02"),
    Venda(3, 101, 502, 1, "2024-09-03"),
    Venda(4, 104, 501, 3, "2024-09-04"),
    Venda(5, 103, 503, 1, "2024-09-05"),
    Venda(6, 102, 502, 1, "2024-09-06"),
]

# --- EXECUÇÃO E DEMONSTRAÇÃO DOS MODELOS LÓGICOS ---

print("--- DEMONSTRAÇÃO DA ARQUITETURA DE DADOS ---")

# 1. Implementação do Data Warehouse (Repositório Geral)
dw_varejo = DataWarehouseVendas()
dw_varejo.carregar_dados_completos(clientes_dados, produtos_dados, vendas_dados)

# Uso do DW para uma métrica corporativa (Modelos Lógicos de Repositórios em ação)
faturamento = dw_varejo.calcular_faturamento_total()
print(f"\n✅ Análise Corporativa (DW): Faturamento Total = R${faturamento:.2f}")

# 2. Implementação do Data Mart (Repositório Especializado para Área de Interesse)
dm_marketing = DataMartMarketing()
dm_marketing.carregar_dados_brutos(clientes_dados, vendas_dados)

# Uso do DM para uma ação de marketing (Área de Interesse em ação)
regiao_foco = "Sul"
clientes_sul = dm_marketing.analisar_clientes_por_regiao(regiao_foco)
print(f"\n🎯 Análise de Marketing (DM): Clientes da Região '{regiao_foco}':")
for c in clientes_sul:
    print(f"- {c.nome}")

# 3. Demonstração de uma Entidade do MLC
print("\n🔍 Entidade Cliente (MLC) Isolada:")
cliente_teste = Cliente(999, "Teste Estrutura", "Teste")
print(cliente_teste)

# A CLASSE Venda representa o RELACIONAMENTO entre Cliente e Produto,
# um pilar do Modelo Lógico Corporativo (MER).

File: semestre-1/disciplina-3-arquitetura-de-dados/unidade-2/aula-2/aula-2.js
// aula-2.js
// Exemplo em JavaScript: Simulação de Modelos Lógicos de Dados para Varejo.
// Usamos classes ES6 para definir as entidades (MLC) e objetos para os repositórios lógicos.

// --- 1. MODELO LÓGICO CORPORATIVO (MLC): As Entidades ---

/**
 * Representa a entidade Cliente (MLC).
 * Define a estrutura lógica dos dados do cliente.
 */
class Cliente {
    constructor(idCliente, nome, regiao) {
        this.idCliente = idCliente; // Atributo: Chave Primária
        this.nome = nome;           // Atributo: Detalhe
        this.regiao = regiao;       // Atributo: Detalhe
    }
}

/**
 * Representa a entidade Produto (MLC).
 */
class Produto {
    constructor(idProduto, nome, preco) {
        this.idProduto = idProduto;
        this.nome = nome;
        this.preco = preco;
    }
}

/**
 * Representa a entidade Venda (MLC), essencialmente o relacionamento entre Cliente e Produto.
 */
class Venda {
    constructor(idVenda, idCliente, idProduto, quantidade, data) {
        this.idVenda = idVenda;
        this.idCliente = idCliente; // Chave Estrangeira (Relacionamento com Cliente)
        this.idProduto = idProduto; // Chave Estrangeira (Relacionamento com Produto)
        this.quantidade = quantidade;
        this.data = data;
    }
}

// --- 2. MODELO DE ÁREA DE INTERESSE (MAI) & MODELO LÓGICO DE REPOSITÓRIO (MLR) ---

/**
 * Simula um Data Mart para a Área de Interesse de Marketing (MLR especializado).
 * Focado em segmentação de clientes.
 */
class DataMartMarketing {
    constructor() {
        // O repositório armazena os dados filtrados e relevantes
        this.clientes = [];
        console.log("\n[MAI/MLR] Data Mart de Marketing inicializado.");
    }

    carregarClientes(dadosClientes) {
        // Simula o processo ETL (Extração, Transformação, Carga) para o Data Mart.
        this.clientes = dadosClientes;
        console.log(`[Marketing] Carregados ${this.clientes.length} clientes para o Data Mart.`);
    }

    analisarClientesPorRegiao(regiaoFoco) {
        /** Exemplo de análise feita pelo departamento de Marketing. */
        return this.clientes.filter(cliente => cliente.regiao === regiaoFoco);
    }
}

/**
 * Simula um Data Warehouse centralizado (MLR abrangente) para análises corporativas.
 * Armazena Fatos (Vendas) e Dimensões (Clientes, Produtos).
 */
class DataWarehouseVendas {
    constructor() {
        this.dimClientes = {};
        this.dimProdutos = {};
        this.fatoVendas = [];
        console.log("\n[MAI/MLR] Data Warehouse de Vendas inicializado.");
    }

    carregarDadosCompletos(clientes, produtos, vendas) {
        // Mapeia Clientes e Produtos por ID para fácil acesso (simulando tabelas de dimensão)
        clientes.forEach(c => this.dimClientes[c.idCliente] = c);
        produtos.forEach(p => this.dimProdutos[p.idProduto] = p);
        
        // Carrega o Fato (Vendas)
        this.fatoVendas = vendas;
        console.log(`[DW Vendas] Carregados ${this.fatoVendas.length} registros de vendas para análise.`);
    }

    calcularFaturamentoTotal() {
        /** Exemplo de métrica de negócio que usa os relacionamentos do modelo lógico. */
        let faturamento = 0;
        
        this.fatoVendas.forEach(venda => {
            const produto = this.dimProdutos[venda.idProduto];
            if (produto) {
                faturamento += venda.quantidade * produto.preco;
            }
        });
        return faturamento;
    }
}

// --- DADOS DE EXEMPLO (População das Entidades MLC) ---

const clientesDados = [
    new Cliente(101, "Ana Silva", "Sul"),
    new Cliente(102, "Bruno Costa", "Norte"),
    new Cliente(103, "Carla Meirelles", "Sul"),
    new Cliente(104, "David Lopes", "Norte"),
];

const produtosDados = [
    new Produto(501, "Celular X", 1500.00),
    new Produto(502, "Notebook Y", 4500.00),
    new Produto(503, "Fone Z", 250.00),
];

const vendasDados = [
    new Venda(1, 101, 501, 1, "2024-09-01"),
    new Venda(2, 102, 503, 2, "2024-09-02"),
    new Venda(3, 101, 502, 1, "2024-09-03"),
    new Venda(4, 104, 501, 3, "2024-09-04"),
    new Venda(5, 103, 503, 1, "2024-09-05"),
    new Venda(6, 102, 502, 1, "2024-09-06"),
];

// --- EXECUÇÃO E DEMONSTRAÇÃO ---

console.log("--- DEMONSTRAÇÃO DA ARQUITETURA DE DADOS EM JAVASCRIPT ---");

// 1. Uso do Data Warehouse (MLR Corporativo)
const dwVarejo = new DataWarehouseVendas();
dwVarejo.carregarDadosCompletos(clientesDados, produtosDados, vendasDados);

const faturamentoTotal = dwVarejo.calcularFaturamentoTotal();
console.log(`\n✅ Análise Corporativa (DW): Faturamento Total = R$${faturamentoTotal.toFixed(2)}`);

// 2. Uso do Data Mart (MLR/MAI Especializado)
const dmMkt = new DataMartMarketing();
// O Data Mart carrega apenas a informação que lhe é relevante
dmMkt.carregarClientes(clientesDados);

const regiaoFoco = "Sul";
const clientesSul = dmMkt.analisarClientesPorRegiao(regiaoFoco);
console.log(`\n🎯 Análise de Marketing (DM): Clientes da Região '${regiaoFoco}':`);
clientesSul.forEach(c => console.log(`- ${c.nome}`));

console.log("\n🔍 Exemplo de Entidade do MLC:");
console.log(new Produto(1000, "Cadeira Ergonômica", 899.99));
// Note que as classes (MLC) fornecem a base estrutural para todos os repositórios.